examples/chip-tool/commands/common/HexConversion.h

```diff
@@ -0,0 +1,65 @@
+/*
+ *   Copyright (c) 2022 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPError.h>
+#include <lib/support/BytesToHex.h>
+#include <lib/support/Span.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+/**
+ * Utility for converting a hex string to bytes, with the right error checking
+ * and allocation size computation.
+ *
+ * Takes a functor to allocate the buffer to use for the hex bytes.  The functor
+ * is expected to return uint8_t *.  The caller is responsible for cleaning up
+ * this buffer as needed.
+ *
+ * On success, *octetCount is filled with the number of octets placed in the
+ * buffer.  On failure, the value of *octetCount is undefined.
+ */
+template <typename F>
+CHIP_ERROR HexToBytes(chip::CharSpan hex, F bufferAllocator, size_t * octetCount)
+{
+    *octetCount = 0;
+
+    if (hex.size() % 2 != 0)
+    {
+        ChipLogError(chipTool, "Error while encoding '%.*s' as an octet string: Odd number of characters.",
+                     static_cast<int>(hex.size()), hex.data());
+        return CHIP_ERROR_INVALID_STRING_LENGTH;
+    }
+
+    const size_t bufferSize = hex.size() / 2;
+    uint8_t * buffer        = bufferAllocator(bufferSize);
+    if (buffer == nullptr)
```

Comment:
When using POSIX `malloc()`/`calloc()` function with 0 length we might get NULL instead of zero-length memory chunk (`chip::Platform::MemoryCalloc` uses POSIX `calloc()`). So, passing zero-length span might fail here, where in fact no buffer is needed, because no decoding will take place. Also, why bufferAllocator functor? Isn't it easier to pass preallocated buffer here? The ratio between bytes and hex-string is always 1:2, so buffer can be preallocated beforehand.

Reply:
Good catch on malloc(0) being allowed to return null.  Will fix.

> Isn't it easier to pass preallocated buffer here?

That requires all callsites to duplicate the length-check and error-handling for incorrect length.

Reply:
https://github.com/project-chip/connectedhomeip/pull/24097/commits/90aa96f5c918ed8b2adc8d8bce0ff96958021b47 should fix the `malloc(0)`/`calloc(0)` bits.